File: 2022/Day9/javascript/Day9.js
// Read File ../input.txt

// Part 1 and Part 2


// ZhengLinLei

const fs = require('fs');

// Read File Async
fs.readFile('../input.txt', 'utf8', (err, text) =>  {
    let steps = {}, visited = new Set(), command = text.split('\n').map(x => x.split(/\s/)).map(x => [x[0], parseInt(x[1])]); command.forEach(x => (steps[x[0]]) ? steps[x[0]] += x[1] : steps[x[0]] = x[1]);

    const move_ = (command) => {
        let posArr = [[0, 0]];
        command.forEach(x => {
            for(let j = x[1]; j > 0; j--){
                let pos = [...posArr.at(-1)]
                switch (x[0]) {
                    case "L": pos[0] -= 1; break;
                    case "R": pos[0] += 1; break;
                    case "U": pos[1] += 1; break;
                    case "D": pos[1] -= 1; break;
                }
                posArr.push(pos);
            }
        });

        return posArr;
    };
    const follow_ = (points) => {
        let posArr = [[0, 0]];
        points.forEach((x, i) => {
            if(i < points.length - 2){
                if(Math.sqrt((points[i+1][0] - posArr[i][0])**2 + (points[i+1][1] - posArr[i][1])**2) > 1){
                    posArr.push([...x]);
                }
            }
        });

        return posArr;
    };

    console.log(follow_(move_(command)))
});

File: 2022/Day9/python/Day9.py
# Part 1 and 2

# Zheng Lin Lei

def _moves(lines):
    current_x, current_y = 0, 0
    yield current_x, current_y
    for line in lines:
        direction, count = line.split()
        for _ in range(int(count)):
            match direction:
                case "L":
                    current_x -= 1
                case "R":
                    current_x += 1
                case "U":
                    current_y -= 1
                case "D":
                    current_y += 1
            yield current_x, current_y


def _follow(points):
    tail_x, tail_y = 0, 0
    yield tail_x, tail_y
    for head_x, head_y in points:
        delta_x, delta_y = head_x - tail_x, head_y - tail_y
        if abs(delta_x) <= 1 and abs(delta_y) <= 1:
            continue
        if abs(delta_x) < abs(delta_y):
            tail_x, tail_y = head_x, head_y - delta_y // abs(delta_y)
        elif abs(delta_x) > abs(delta_y):
            tail_x, tail_y = head_x - delta_x // abs(delta_x), head_y
        else:
            tail_x, tail_y = head_x - delta_x // abs(delta_x), head_y - delta_y // abs(
                delta_y
            )
        yield tail_x, tail_y


lines = open("../input.txt").read().splitlines()

# print(list(_moves(lines)))
print("Part 1:", len(set(_follow(_moves(lines)))))
print("Part 2:", len(set(_follow(_follow(_follow(_follow(_follow(_follow(_follow(_follow(_follow(_moves(lines)))))))))))))